[VL] RAS: Add a new built-in cost model that avoids offloading trivial projects if its neighbor nodes fell back

zml1206 opened the pull request:
## What changes were proposed in this pull request?

1. Rename `RoughCostModel` to `LegacyCostModel`, set default cost model as legacy.
2. Add new `RoughCostModel`, make projectList all cheap `ProjectExec` has the same cost as `ProjectExecTransform`, to reduce unnecessary c2r and r2c for avoid performance regressions.

## How was this patch tested?
UT

zhztheplayer commented:
Thanks. I've retitled the PR to make things clearer.

zhztheplayer commented on gluten-core/src/main/scala/org/apache/gluten/planner/cost/RoughCostModel.scala:
```diff
@@ -0,0 +1,50 @@
+import org.apache.spark.sql.catalyst.expressions.{Alias, Attribute, NamedExpression}
+import org.apache.spark.sql.execution.{ColumnarToRowExec, ProjectExec, RowToColumnarExec, SparkPlan}
+
+class RoughCostModel extends LegacyCostModel {
```
Let's add a base class `abstract class LongCostModel` then make the two implementations inherit from it.

zhztheplayer commented on gluten-core/src/main/scala/org/apache/gluten/planner/cost/RoughCostModel.scala:
```diff
@@ -0,0 +1,50 @@
+      case _: RemoveFilter.NoopFilter =>
+        // To make planner choose the tree that has applied rule PushFilterToScan.
+        0L
+      case ProjectExec(projectList, _) if projectList.forall(isCheapExpression) => 10L
```
Could add a short comment here to explain about the coster.

zhztheplayer reviewed (commented):
And would you like to add a test suite for `RoughCostModel`? Could be placed as `org.apache.gluten.execution.VeloxRoughCostModelSuite`

zml1206 commented:
This UT is very strange. My local verification was normal, but it failed on github. I added logs later in #6760 , output the plan, and found that the scan node was still offloaded, but I added the configuration GlutenConfig.COLUMNAR_FILESCAN_ENABLED.key -> "false" .
```
+- ^(4230) ProjectExecTransformer [c1#80388 AS c3#80394]
   +- ^(4230) NativeFileScan parquet spark_catalog.default.tmp1[c1#80388] Batched: true, DataFilters: [], Format: Parquet, Location: InMemoryFileIndex(1 paths)[file:/__w/incubator-gluten/incubator- gluten/backends-velox/spark-wareh..., PartitionFilters: [], PushedFilters: [], ReadSchema: struct<c1:int>
```
cc @zhztheplayer Do you know the reason? Thank you.

zhztheplayer commented:
@zml1206 

It's because of a bug that makes configuration changes to RAS planner not take effect. https://github.com/apache/incubator-gluten/pull/6766 should fix the issue. Thanks.

zhztheplayer reviewed (approved):
Thanks!

zhztheplayer commented on shims/common/src/main/scala/org/apache/gluten/GlutenConfig.scala:
```diff
@@ -1320,7 +1320,7 @@ object GlutenConfig {
         "Experimental: The class name of user-defined cost model that will be used by RAS. " +
           "If not specified, a rough built-in cost model will be used.")
```
nit: "If not specified, a legacy built-in cost model that exhaustively offloads computations will be used"

zhztheplayer commented on backends-velox/src/test/scala/org/apache/gluten/execution/VeloxRoughCostModelSuite.scala:
```diff
@@ -0,0 +1,54 @@
+
+  test("fallback trivial project if its neighbor nodes fell back") {
+    withSQLConf(GlutenConfig.COLUMNAR_FILESCAN_ENABLED.key -> "false") {
+      runQueryAndCompare("select c1 as c3 from tmp1") {
```
can we simply use `select c1 from tmp1`?

zml1206 replied:
`select c1 from tmp1` will remove project.

zhztheplayer reviewed (commented):
Just some minors left. Thanks.
